sw/device/lib/dif/dif_pwm.c

```diff
@@ -0,0 +1,157 @@
+    return kDifLocked;
+  }
+
+  // Configure duty cycle register.
+
+  // Get "beats_per_pulse_cycle" from the PWM config register.
+  // There are 2 ^ (`duty_cycle_resolution` + 1) "beats" per "pulse cycle".
+  uint8_t duty_cycle_resolution = bitfield_field32_read(
+      mmio_region_read32(pwm->base_addr, PWM_CFG_REG_OFFSET),
+      PWM_CFG_DC_RESN_FIELD);
+  uint32_t beats_per_pulse_cycle = 1U << (duty_cycle_resolution + 1);
+
+  // Check duty cycle configurations.
+  if (config.duty_cycle_a >= beats_per_pulse_cycle ||
+      config.duty_cycle_b >= beats_per_pulse_cycle) {
+    return kDifBadArg;
+  }
+
+  // There are 2 ^ 16 "phase counter ticks" in one "pulse cycle", and therefore
+  // 2 ^ (16 - `duty_cycle_resolution` - 1) "phase counter ticks" in one "beat".
+  uint16_t phase_cntr_ticks_per_beat = 1U << (16 - duty_cycle_resolution - 1);
+  uint32_t duty_cycle_reg =
+      bitfield_field32_write(0, PWM_DUTY_CYCLE_0_A_0_FIELD,
+                             phase_cntr_ticks_per_beat * config.duty_cycle_a);
+  duty_cycle_reg =
+      bitfield_field32_write(duty_cycle_reg, PWM_DUTY_CYCLE_0_B_0_FIELD,
+                             phase_cntr_ticks_per_beat * config.duty_cycle_b);
+
+  // Configure parameter register.
+  uint32_t param_reg = bitfield_field32_write(
+      0, PWM_PWM_PARAM_0_PHASE_DELAY_0_FIELD, config.phase_delay);
+  if (config.mode == kDifPwmModeHeartbeat) {
+    param_reg =
+        bitfield_bit32_write(param_reg, PWM_PWM_PARAM_0_HTBT_EN_0_BIT, true);
+  } else if (config.mode == kDifPwmModeBlink) {
+    param_reg =
+        bitfield_bit32_write(param_reg, PWM_PWM_PARAM_0_BLINK_EN_0_BIT, true);
+  }
+
+  // Configure polarity register.
+  uint32_t invert_reg =
+      mmio_region_read32(pwm->base_addr, PWM_INVERT_REG_OFFSET);
+
+  // Configure blink mode parameter register.
+  uint32_t blink_param_reg = 0;
+  if (config.mode == kDifPwmModeHeartbeat || config.mode == kDifPwmModeBlink) {
+    blink_param_reg = bitfield_field32_write(
+        blink_param_reg, PWM_BLINK_PARAM_0_X_0_FIELD, config.blink_parameter_x);
+    if (config.mode == kDifPwmModeHeartbeat) {
+      if (config.blink_parameter_y >= beats_per_pulse_cycle) {
+        return kDifBadArg;
+      }
+      // Convert "beats" to "phase counter ticks", since this value is added to
+      // the duty cycle (which hardware computes in "phase counter ticks").
+      blink_param_reg = bitfield_field32_write(
+          blink_param_reg, PWM_BLINK_PARAM_0_Y_0_FIELD,
+          phase_cntr_ticks_per_beat * config.blink_parameter_y);
+    } else if (config.mode == kDifPwmModeBlink) {
+      blink_param_reg =
+          bitfield_field32_write(blink_param_reg, PWM_BLINK_PARAM_0_Y_0_FIELD,
+                                 config.blink_parameter_y);
+    }
+  } else if (config.mode != kDifPwmModeFirmware) {
+    return kDifBadArg;
+  }
+
+#define PWM_CHANNEL_CONFIG_CASE_(channel_)                                     \
```

Comment:
A separate inline function might be a little nicer than stuffing the logic in the macro, but that's up to you. This comes down to style preferences, and what you have is readable.

Reply:
I'm not sure if that will clean this up too much, because the inline function would need to take as input 4 different register offsets (since the X-macro setup is doing some code gen), and additionally all the configuration settings. Unless I'm missing something?

Reply:
Yeah, it's just style preference: I do not personally prefer to see much happening in a macro. If you like this macro style better, though, carry on :)